make/langtools/tools/javacserver/client/Client.java

```diff
@@ -0,0 +1,190 @@
+        int exitCode = client.dispatchToServer();
+
+        System.exit(exitCode);
+    }
+
+    private int dispatchToServer() {
+        try {
+            // Check if server seems to be already running
+            if (!conf.portFile().hasValidValues()) {
+                // Fork a new server and wait for it to start
+                startNewServer();
+            }
+
+            try (Socket socket = tryConnect()) {
+                BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+                PrintWriter out = new PrintWriter(new OutputStreamWriter(socket.getOutputStream()));
+
+                Protocol.sendCommand(out, conf.javacArgs());
+                int exitCode = Protocol.readResponse(in);
+
+                return exitCode;
+            }
+        } catch (PortFileInaccessibleException e) {
+            Log.error("Port file inaccessible.");
+            return Result.ERROR.exitCode;
+        } catch (IOException ioe) {
+            Log.error("IOException caught during compilation: " + ioe.getMessage());
+            Log.debug(ioe);
+            return Result.ERROR.exitCode;
+        } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt(); // Restore interrupt
+            Log.error("Compilation interrupted.");
+            Log.debug(ie);
+            return Result.ERROR.exitCode;
+        }
+    }
+
+    /*
+     * Makes MAX_CONNECT_ATTEMPTS attempts to connect to server.
+     */
+    private Socket tryConnect() throws IOException, InterruptedException {
+        int attempt = 0;
+
+        while (true) {
+            Log.debug("Trying to connect. Attempt " + (++attempt) + " of " + MAX_CONNECT_ATTEMPTS);
+            try {
+                Socket socket = new Socket();
+                InetAddress localhost = InetAddress.getByName(null);
+                InetSocketAddress address = new InetSocketAddress(localhost, conf.portFile().getPort());
+                socket.connect(address, CONNECTION_TIMEOUT);
+                Log.debug("Connected");
+                return socket;
+            } catch (IOException ex) {
+                Log.error("Connection attempt failed: " + ex.getMessage());
+                if (attempt >= MAX_CONNECT_ATTEMPTS) {
+                    Log.error("Giving up");
+                    throw new IOException("Could not connect to server", ex);
+                }
+            }
+            Thread.sleep(WAIT_BETWEEN_CONNECT_ATTEMPTS);
+        }
+    }
+
+    /*
+     * Fork a server process process and wait for server to come around
+     */
+    private void startNewServer() throws IOException, InterruptedException {
+        List<String> cmd = new ArrayList<>();
+        // conf.javaCommand() is how to start java in the way we want to run
+        // the server
+        cmd.addAll(Arrays.asList(conf.javaCommand().split(" ")));
+        // javacserver.server.Server is the server main class
+        cmd.add("javacserver.server.Server");
```

Comment:
No, can't say I have. Do you think there is anything to gain by doing that? (Apart from possibly getting this right automatically if the class is renamed by IDE tools) (I'm not sure it would be a win in readability)

Reply:
I like to reference class names that way so that it's typed instead of just a string. If the class changes name, you get a compilation error, and the IDE would handle refactoring automatically. Not a strong need in this particular instance, just something I generally tend to do.